bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+                self.read_buf[..FRAME_HEADER_SIZE].try_into().unwrap(),
+            );
+            let end = size + FRAME_HEADER_SIZE;
+            if end > self.total_read {
+                return Ok(());
+            }
+            let msg: Msg = match ciborium::from_reader(
+                &self.read_buf[FRAME_HEADER_SIZE..end],
+            ) {
+                Ok(msg) => {
+                    // Move any remaining bytes to the beginning of the buffer.
+                    self.read_buf.copy_within(end..self.total_read, 0);
+                    self.total_read = self.total_read - end;
+                    msg
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: failed to deserialize: {}", e
+                    );
+                    return self.close().await;
+                }
+            };
+            self.last_received_msg = Instant::now();
+            debug!(self.log, "Received {:?}", msg);
+            if let Msg::Fsm(msg) = msg {
+                if let Err(e) = self
+                    .main_tx
+                    .send(ConnToMainMsg {
+                        handle_unique_id: self.unique_id,
+                        msg: ConnToMainMsgInner::Received {
+                            from: self.peer_id.clone(),
+                            msg,
+                        },
+                    })
+                    .await
+                {
+                    warn!(
+                        self.log,
+                        "Failed to send received msg to main task: {:?}", e
+                    );
+                }
+            }
+        }
+    }
+
+    // Send ping messages and check for inactivity timeouts
+    async fn ping(&mut self) -> Result<(), ConnErr> {
+        if Instant::now() - self.last_received_msg > INACTIVITY_TIMEOUT {
+            warn!(self.log, "Closing connection: inactivity timeout",);
+            return self.close().await;
+        }
+        if self.write_queue.len() == MSG_WRITE_QUEUE_CAPACITY {
+            warn!(self.log, "Closing connection: write queue full",);
+            self.close().await
+        } else {
+            match write_framed(&Msg::Ping) {
+                Ok(msg) => {
+                    self.write_queue.push_back(msg);
+                    Ok(())
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: Failed to serialize msg: {}", e
+                    );
+                    self.close().await
+                }
+            }
+        }
+    }
+
+    // Close and drop the connection.
+    async fn close(&mut self) -> Result<(), ConnErr> {
+        if let Err(e) = self
+            .main_tx
+            .send(ConnToMainMsg {
+                handle_unique_id: self.unique_id,
+                msg: ConnToMainMsgInner::Disconnected {
```

Comment:
Is it weird that if the main task sends us `MainToConnMsg::Close`, we both reply along the `main_tx` channel with `Disconnected` **and** return `Err(ConnErr::Close)`?

Reply:
Somewhat. But that's not the only reason this method is called. The connection can close itself and we must inform the main task so it can clean up.

Reply:
Yeah, and reading further, it looks like the return code from `run` is either discarded (for a server task) or used to decide whether to restart (for a client task), so it doesn't make it back to the main task either way.